satpy/modifiers/parallax.py

```diff
@@ -0,0 +1,503 @@
+from satpy.utils import get_satpos, lonlat2xyz, xyz2lonlat
+
+logger = logging.getLogger(__name__)
+
+
+class MissingHeightError(ValueError):
+    """Raised when heights do not overlap with area to be corrected."""
+
+
+class IncompleteHeightWarning(UserWarning):
+    """Raised when heights only partially overlap with area to be corrected."""
+
+
+def forward_parallax(sat_lon, sat_lat, sat_alt, lon, lat, height):
+    """Calculate forward parallax effect.
+
+    Calculate the forward parallax effect.  When a satellite instrument
+    observes the Earth, the geolocation assumes it sees the Earth surface
+    at the geoid (elevation zero).  In reality, the field of view may
+    stop short of the geoid as it observes, for example, a cloud or
+    elevated ground.  This function calculates the forward parallax
+    effect.  If the view of a pixel at location (lat, lon) is blocked
+    by a cloud at height h, we calculate the location of this blocking.
+
+    Calculate parallax correction based on satellite position and
+    (cloud top) height coordinates in geodetic (unprojected) coordinates.
+    This function calculates the latitude and longitude belonging to the
+    cloud top, based on the location of the satellite and the location
+    of the cloud.
+
+    For scenes that are only partly cloudy, the user might set the cloud top
+    height for clear-sky pixels to NaN.  This function will return a corrected
+    lat/lon as NaN as well.  The user can use the original lat/lon for those
+    pixels or use the higher level :class:`ParallaxCorrection` class.
+
+    This function assumes a spherical Earth.
+
+    Args:
+        sat_lon (number): Satellite longitude in geodetic coordinates [°]
+        sat_lat (number): Satellite latitude in geodetic coordinates [°]
+        sat_alt (number): Satellite altitude above the Earth surface [m]
+        lon (array or number): Longitudes of pixel or pixels to be corrected,
+            in geodetic coordinates [°]
+        lat (array or number): Latitudes of pixel/pixels to be corrected, in
+            geodetic coordinates [°]
+        height (array or number): Heights of pixels on which the correction
+            will be based.  Typically this is the cloud top height. [m]
+
+    Returns:
+        tuple[float, float]: New geolocation
+            New geolocation ``(lon, lat)`` for the longitude and
+            latitude that were to be corrected, in geodetic coordinates. [°]
+    """
+    # Be careful with units here.  Heights may be either in m or km, and may
+    # refer to either the Earth's surface on the Earth's centre.  Cloud top
+    # height is usually reported in metres above the Earth's surface, rarely in
+    # km.  Satellite altitude may be reported in either m or km, but orbital
+    # parameters may be in relation the the Earths centre.  The Earth radius
+    # from pyresample is reported in km.
+
+    elevation = _get_satellite_elevation(sat_lon, sat_lat, sat_alt, lon, lat)
```

Comment:
How about the cases where the satellite viewing angles are available from the dataset itself? could we use that as an optional dependency for the modifier?

Reply:
Perhaps.  But one would still have to calculate the satellite elevation from the satellite viewing angle (Earth curvature).  Like in my previous comment (responding to your comment on lines 274-277), we'd need to think about the API.

Reply:
according to this https://github.com/pytroll/satpy/blob/4971b7cc49ecee9bfeeb1894ef38a3e3b02cb6f7/satpy/modifiers/angles.py#L445
the elevation is 90 - satellite_zenith_angle...

Reply:
I thought the satellite zenith angle was defined from the satellite POV, in which case this relation would only hold for a flat Earth.  But maybe I'm wrong about the definition of satellite zenith angle.

Reply:
I agree with @gerritholl. I'm familiar with satellite zenith being defined so that it is zero degrees when the satellite is directly overhead a pixel. So for SEVIRI, zenith would be zero degrees at the subsatellite point and increase as you move towards the edge of the disk.

Reply:
I'm deferring getting the viewing angle from the dataset itself to a future PR.  I have added this to the main issue description and will create corresponding issues after this PR has been merged.